include/swift/Runtime/Backtrace.h

```diff
@@ -123,8 +134,42 @@ struct BacktraceSettings {
 
 SWIFT_RUNTIME_STDLIB_INTERNAL BacktraceSettings _swift_backtraceSettings;
 
-SWIFT_RUNTIME_STDLIB_SPI SWIFT_CC(swift) bool _swift_isThunkFunction(const char *mangledName);
-
+SWIFT_RUNTIME_STDLIB_SPI
+bool _swift_backtrace_isThunkFunction(const char *mangledName);
+
+/// Try to demangle a symbol.
+///
+/// Unlike other entry points that do this, we try both Swift and C++ here.
+///
+/// @param mangledName is the symbol name to be demangled.
+/// @param mangledNameLength is the length of this name.
+/// @param outputBuffer is a pointer to a buffer in which to place the result.
+/// @param outputBufferSize points to a variable that contains the size of the
+/// output buffer.
+/// @param status returns the status codes defined in the C++ ABI.
+///
+/// If outputBuffer is nullptr, the function will allocate memory for the
+/// result using malloc().  In this case, outputBufferSize may be nullptr;
+/// if it is *not* nullptr, it will be set to the size of buffer that was
+/// allocated.  This is not necessarily the length of the string (it may be
+/// somewhat higher).
+///
+/// Otherwise, the result will be written into the output buffer, and the
+/// size of the result will be written into outputBufferSize.  If the buffer
+/// is too small, the result will be truncated, but outputBufferSize will
+/// still be set to the number of bytes that would have been required to
+/// copy out the full result (including a trailing NUL).
+///
+/// The unusual behaviour here is a consequence of the way the C++ ABI's
+/// demangling function works.
+///
+/// @returns a pointer to the output if demangling was successful.
+SWIFT_RUNTIME_STDLIB_SPI
+char *_swift_backtrace_demangle(const char *mangledName,
```

Comment:
Do we want to indicate whether it was a swift symbol or C++ symbol that was demangled? The status is a little weird otherwise? And do we want to document the values of status?

e.g. (no idea what Windows does here though so... ¯\\\_(ツ)\_/¯)

```
*status is set to one of the values:
 0: The demangling operation succeeded.
-1: A memory allocation failiure occurred.
-2: mangled_name is not a valid name under the C++ ABI mangling rules.
-3: One of the arguments is invalid.
```

Reply:
(I guess the `-2` value should read that the `mangled_name` is not valid under the C++ ABI or Swift mangling rules)

Reply:
Windows is totally different, helpfully.

Reply:
😒 Windows... Are we going to be internally consistent and always returning the same set of status numbers for the various kinds of failures, or are we going to be externally consistent and return... a number?

Reply:
I'm not too bothered about this, as it's SPI that's only used by the backtracing implementation (which is built in lock-step with the runtime and right now actually ignores `status` anyway — it's really only there for debugging; if we can't demangle something, we just don't), and I don't even mind the signature changing between Windows and Linux. macOS isn't using this function.